### DISCUSSION TITLE
NotImplementedError: Module is missing the required "forward" function

### DISCUSSION THREAD

**Discussion by Sarthak-Singh04** [mrdbourke/pytorch-deep-learning#97] [Q&A] [answered]
![Screenshot (40)](https://user-images.githubusercontent.com/106304584/184190341-23ab9446-ba91-45ad-8f56-b57a4ce2ec3f.png)
![Screenshot (41)](https://user-images.githubusercontent.com/106304584/184190354-56324cb1-dfe3-448c-b30a-a134d5b2b3b2.png)

I don't know why they show error in this code I tried lot of thing it didn't found it what is problem in this code

**Comment by Pouyaexe** [comment, marked as answer]
Well, you are defining the forward function inside the def __init__ function. You should remove the indentation before the def forward.
I recommend you to turn on the Indentation guides from tools>settings>editor. 
Example of a correct implementation: 
<img width="777" alt="image" src="https://user-images.githubusercontent.com/40316639/184352032-f29c5a95-bac7-41dc-9409-6009d60b0f5a.png">

**Comment by mrdbourke** [comment]
Hi @Sarthakparmar-04,

@Pouyaexe is right, the issue is with your indentation of the `forward()` function.

Notice the `forward()` method in the code below is inline with the `def __init__()` method: 

```python
from torch import nn

# Build model
class BlobModel(nn.Module):
    def __init__(self, input_features, output_features, hidden_units=8):
        """Initializes all required hyperparameters for a multi-class classification model.

        Args:
            input_features (int): Number of input features to the model.
            out_features (int): Number of output features of the model
              (how many classes there are).
            hidden_units (int): Number of hidden units between layers, default 8.
        """
        super().__init__()
        self.linear_layer_stack = nn.Sequential(
            nn.Linear(in_features=input_features, out_features=hidden_units),
            # nn.ReLU(), # <- does our dataset require non-linear layers? (try uncommenting and see if the results change)
            nn.Linear(in_features=hidden_units, out_features=hidden_units),
            # nn.ReLU(), # <- does our dataset require non-linear layers? (try uncommenting and see if the results change)
            nn.Linear(in_features=hidden_units, out_features=output_features), # how many classes are there?
        )
    
    def forward(self, x):
        return self.linear_layer_stack(x)

# Create an instance of BlobModel and send it to the target device
model_4 = BlobModel(input_features=NUM_FEATURES, 
                    output_features=NUM_CLASSES, 
                    hidden_units=8).to(device)
model_4
```

Without the `forward()` method being inline with `def __init__()`, the code will error. 

As @Pouyaexe said, you can turn on indentation guides in Google Colab by going to tools -> settings -> editor -> show indentation guides.

Source: https://www.learnpytorch.io/02_pytorch_classification/#82-building-a-multi-class-classification-model-in-pytorch

**Reply by Pouyaexe** [reply]
Thanks for the example, Daniel! Keep up the good work

**Reply by Sarthak-Singh04** [reply]
yeah bro I understand what's the problem in the code It's silly mistake I did in the code

**Reply by mrdbourke** [reply]
Don’t worry @Sarthakparmar-04, I’ve made that error 100 times… that’s the only reason I know the fix haha

**Comment by sbeale007** [comment]
Hi! I am having this same problem when I try to test my model. This is my code for loading my model and trying to predict 

`import mlflow
logged_model = ''

loaded_model = mlflow.pyfunc.load_model(logged_model)

x = torch.load("/Volumes/LaCie/smallest_nonan_ml2/test/uas/lr/uas_0.pt")
px = pd.DataFrame(x)

loaded_model.predict(px)`

and I get the error 

<img width="664" alt="Screen Shot 2023-06-15 at 10 43 11 AM" src="https://github.com/mrdbourke/pytorch-deep-learning/assets/132512638/a56418ca-2fef-4ed7-a581-95ecde55fdec">

I have checked my indentation as in this post and my forward function is in line with my __init__(). Is there another fix to this? 

Thank you!
